File: src/dotnet/Signalbird.Sdk/PartnerClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;

namespace Signalbird.Sdk;

/// <summary>
/// Partner istemcisinin kurulumu. Anahtar <c>sbp_live_…</c>'dir; takım
/// anahtarı (<c>sb_…</c>) burada çalışmaz.
/// </summary>
public sealed class SignalbirdPartnerOptions
{
    /// <summary>Sözleşmeli partner anahtarı (<c>sbp_live_…</c>). Tarayıcıya İNMEZ.</summary>
    public string ApiKey { get; set; } = string.Empty;

    public string BaseUrl { get; set; } = "https://signalbird.io/api";

    public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(15);

    public bool ThrowOnError { get; set; }
}

/// <summary>
/// Partner istemcisi — BEŞİNCİ yüzey.
///
/// <para>
/// Signalbird'ü kendi ürününün içinde satan sözleşmeli platform (veribenim,
/// submitcms) müşterisini bununla sağlar ve yetkilendirir.
/// </para>
///
/// <para>
/// Bu, "Admin yüzeyi OLMAYACAK" kuralının BİLİNÇLİ istisnasıdır ve istisna
/// olduğu için ayrı anahtar türü taşır. Kural, müşterinin kendi anahtarıyla
/// (<c>sb_…</c>) şirket açamaması içindi; o kural aynen duruyor.
/// </para>
///
/// <para>
/// Partner SÜPER YÖNETİCİ DEĞİLDİR: yalnız kendi açtığı company'lere erişir,
/// başkasınınki 404 döner.
/// </para>
///
/// <para>Sözleşme: docs/CONTRACT.md § 12</para>
/// </summary>
public sealed class PartnerClient
{
    private readonly Transport _http;

    public PartnerClient(SignalbirdPartnerOptions options, HttpClient? http = null)
    {
        if (string.IsNullOrEmpty(options.ApiKey))
        {
            throw new SignalbirdException("Signalbird: ApiKey zorunlu.", 0, "NO_KEY");
        }

        if (!options.ApiKey.StartsWith("sbp_live_", StringComparison.Ordinal))
        {
            throw new SignalbirdException(
                "Signalbird: partner istemcisi partner anahtarı ister (sbp_live_…). "
                + "Takım (sb_…), Telsiz (sbr_…) ve uygulama (sbw_pub_…) anahtarları burada çalışmaz.",
                0,
                "WRONG_KEY_TYPE");
        }

        _http = new Transport(options.ApiKey, options.BaseUrl, options.Timeout, options.ThrowOnError, http);
    }

    // ── Müşteri ───────────────────────────────────────────────────────────

    /// <summary>
    /// Company + takım + owner açar. IDEMPOTENTTİR: aynı <c>external_id</c> ile
    /// ikinci çağrı yeni kayıt açmaz. <c>keys</c> yalnız ilk oluşturmada döner.
    /// </summary>
    public Task<SbResult> CreateCompanyAsync(object input, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Post, "/v1/partner/companies", input, null, ct);

    public Task<SbResult> ListCompaniesAsync(IDictionary<string, object?>? query = null, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Get, "/v1/partner/companies", null, query, ct);

    public Task<SbResult> GetCompanyAsync(object externalId, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Get, $"/v1/partner/companies/{Transport.Seg(externalId)}", null, null, ct);

    public Task<SbResult> UpdateCompanyAsync(object externalId, object input, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Patch, $"/v1/partner/companies/{Transport.Seg(externalId)}", input, null, ct);

    /// <summary>Askıya alır — SİLMEZ. İzleme ve mesaj geçmişi durur.</summary>
    public Task<SbResult> SuspendCompanyAsync(object externalId, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Delete, $"/v1/partner/companies/{Transport.Seg(externalId)}", null, null, ct);

    public Task<SbResult> RotateKeyAsync(object externalId, string type, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Post, $"/v1/partner/companies/{Transport.Seg(externalId)}/keys/rotate",
            new Dictionary<string, object?> { ["type"] = type }, null, ct);

    // ── Domain ────────────────────────────────────────────────────────────

    /// <summary>
    /// Domain ekler ve (istenirse) izlemeye alır. Kayıt
    /// <c>verified_via='partner'</c> ile doğar: izleme, sohbet ve push için
    /// yeter — e-posta/SMS KAMPANYASI için TXT şarttır.
    /// </summary>
    public Task<SbResult> AddDomainAsync(object companyExternalId, object input, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Post, $"/v1/partner/companies/{Transport.Seg(companyExternalId)}/domains", input, null, ct);

    public Task<SbResult> ListDomainsAsync(object companyExternalId, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Get, $"/v1/partner/companies/{Transport.Seg(companyExternalId)}/domains", null, null, ct);

    public Task<SbResult> GetDomainAsync(object externalId, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Get, $"/v1/partner/domains/{Transport.Seg(externalId)}", null, null, ct);

    public Task<SbResult> VerifyDomainAsync(object externalId, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Post, $"/v1/partner/domains/{Transport.Seg(externalId)}/verify", null, null, ct);

    public Task<SbResult> RemoveDomainAsync(object externalId, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Delete, $"/v1/partner/domains/{Transport.Seg(externalId)}", null, null, ct);

    /// <summary>Hiç kontrol yoksa <c>uptime</c> null döner, %100 DEĞİL.</summary>
    public Task<SbResult> DomainUptimeAsync(object externalId, string range = "24h", CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Get, $"/v1/partner/domains/{Transport.Seg(externalId)}/uptime", null,
            new Dictionary<string, object?> { ["range"] = range }, ct);

    /// <summary>Tek istekte tüm domainler — liste ekranı N+1 atmasın.</summary>
    public Task<SbResult> CompanyUptimeAsync(object companyExternalId, string range = "24h", CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Get, $"/v1/partner/companies/{Transport.Seg(companyExternalId)}/uptime", null,
            new Dictionary<string, object?> { ["range"] = range }, ct);

    // ── Modül yetkisi ─────────────────────────────────────────────────────

    public Task<SbResult> ListModulesAsync(object companyExternalId, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Get, $"/v1/partner/companies/{Transport.Seg(companyExternalId)}/modules", null, null, ct);

    /// <summary>"Bu müşteri şu modül için ödeme yaptı, kullanabilir."</summary>
    public Task<SbResult> GrantModuleAsync(object companyExternalId, object input, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Post, $"/v1/partner/companies/{Transport.Seg(companyExternalId)}/modules", input, null, ct);

    /// <summary>Yalnız partner'ın KENDİ verdiği hakkı geri alır; plan hakkına dokunmaz.</summary>
    public Task<SbResult> RevokeModuleAsync(object companyExternalId, string module, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Delete,
            $"/v1/partner/companies/{Transport.Seg(companyExternalId)}/modules/{Transport.Seg(module)}", null, null, ct);

    // ── Kullanıcı ─────────────────────────────────────────────────────────

    public Task<SbResult> CreateUserAsync(object companyExternalId, object input, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Post, $"/v1/partner/companies/{Transport.Seg(companyExternalId)}/users", input, null, ct);

    public Task<SbResult> ListUsersAsync(object companyExternalId, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Get, $"/v1/partner/companies/{Transport.Seg(companyExternalId)}/users", null, null, ct);

    /// <summary>Üyeliği kaldırır, kişinin Signalbird hesabını SİLMEZ.</summary>
    public Task<SbResult> RemoveUserAsync(object companyExternalId, object userExternalId, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Delete,
            $"/v1/partner/companies/{Transport.Seg(companyExternalId)}/users/{Transport.Seg(userExternalId)}", null, null, ct);

    // ── Gömme ─────────────────────────────────────────────────────────────

    /// <summary>
    /// Panel ekranını partner sayfasına gömmek için kısa ömürlü jeton:
    /// 120 saniye yaşar ve TEK KULLANIMLIKTIR — jeton URL'de gider, log ve
    /// <c>Referer</c> başlığına düşer.
    /// </summary>
    public Task<SbResult> CreateEmbedTokenAsync(object companyExternalId, object input, CancellationToken ct = default)
        => _http.RequestAsync(HttpMethod.Post, $"/v1/partner/companies/{Transport.Seg(companyExternalId)}/embed", input, null, ct);
}

File: src/php/Partner/PartnerClient.php
<?php

namespace Signalbird\Sdk\Partner;

use Signalbird\Sdk\SignalbirdException;

/**
 * Partner istemcisi — BEŞİNCİ yüzey.
 *
 * Signalbird'ü kendi ürününün içinde satan sözleşmeli platform (veribenim,
 * submitcms) müşterisini bununla sağlar ve yetkilendirir.
 *
 * **CLAUDE.md'deki "Admin yüzeyi OLMAYACAK" kuralının bilinçli istisnasıdır**
 * ve istisna olduğu için ayrı anahtar türü taşır (`sbp_live_…`). Kural,
 * müşterinin kendi anahtarıyla (`sb_`) şirket açamaması içindi; o kural aynen
 * duruyor.
 *
 * Node karşılığı: src/node/partner.ts — davranış birebir aynıdır.
 * Sözleşme: docs/CONTRACT.md § 12
 */
class PartnerClient
{
    public const DEFAULT_BASE_URL = 'https://live.signalbird.io/api';

    private string $baseUrl;

    public function __construct(
        private string $apiKey,
        ?string $baseUrl = null,
        private int $timeout = 15,
        private bool $throwOnError = false,
    ) {
        if ($apiKey === '') {
            throw new SignalbirdException('Signalbird: apiKey zorunlu.', 'NO_KEY', 0);
        }

        if (! str_starts_with($apiKey, 'sbp_live_')) {
            throw new SignalbirdException(
                'Signalbird: partner istemcisi partner anahtarı ister (sbp_live_…). '
                . 'Takım (sb_…), Telsiz (sbr_…) ve uygulama (sbw_pub_…) anahtarları burada çalışmaz.',
                'WRONG_KEY_TYPE',
                0,
            );
        }

        $this->baseUrl = rtrim($baseUrl ?: self::DEFAULT_BASE_URL, '/');
    }

    // ── Müşteri ───────────────────────────────────────────────────────────

    /**
     * Company + takım + owner açar. **Idempotenttir**: aynı `external_id` ile
     * ikinci çağrı `created:false` ile var olanı döner ve anahtarları TEKRAR
     * DÖNMEZ.
     *
     * @param array<string, mixed> $input
     */
    public function createCompany(array $input): array
    {
        return $this->request('POST', '/v1/partner/companies', $input);
    }

    /** @param array<string, mixed> $query */
    public function listCompanies(array $query = []): array
    {
        return $this->request('GET', '/v1/partner/companies', null, $query);
    }

    public function getCompany(string $externalId): array
    {
        return $this->request('GET', '/v1/partner/companies/' . rawurlencode($externalId));
    }

    /** @param array<string, mixed> $input */
    public function updateCompany(string $externalId, array $input): array
    {
        return $this->request('PATCH', '/v1/partner/companies/' . rawurlencode($externalId), $input);
    }

    /** Askıya alır — SİLMEZ. */
    public function suspendCompany(string $externalId): array
    {
        return $this->request('DELETE', '/v1/partner/companies/' . rawurlencode($externalId));
    }

    public function rotateKey(string $externalId, string $type): array
    {
        return $this->request(
            'POST',
            '/v1/partner/companies/' . rawurlencode($externalId) . '/keys/rotate',
            ['type' => $type],
        );
    }

    // ── Domain ────────────────────────────────────────────────────────────

    /**
     * Domain ekler ve (istenirse) izlemeye alır. Kayıt `verified_via:'partner'`
     * ile doğar: e-posta/SMS kampanyası için TXT şarttır — yanıttaki `dns`
     * kaydını yayınlayıp `verifyDomain` çağırmak kapıyı açar.
     *
     * @param array<string, mixed> $input
     */
    public function addDomain(string $companyExternalId, array $input): array
    {
        return $this->request(
            'POST',
            '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/domains',
            $input,
        );
    }

    public function listDomains(string $companyExternalId): array
    {
        return $this->request('GET', '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/domains');
    }

    public function getDomain(string $externalId): array
    {
        return $this->request('GET', '/v1/partner/domains/' . rawurlencode($externalId));
    }

    public function verifyDomain(string $externalId): array
    {
        return $this->request('POST', '/v1/partner/domains/' . rawurlencode($externalId) . '/verify');
    }

    public function removeDomain(string $externalId): array
    {
        return $this->request('DELETE', '/v1/partner/domains/' . rawurlencode($externalId));
    }

    public function domainUptime(string $externalId, string $range = '24h'): array
    {
        return $this->request(
            'GET',
            '/v1/partner/domains/' . rawurlencode($externalId) . '/uptime',
            null,
            ['range' => $range],
        );
    }

    /** Tek istekte müşterinin tüm domainleri — liste ekranı N+1 atmasın. */
    public function companyUptime(string $companyExternalId, string $range = '24h'): array
    {
        return $this->request(
            'GET',
            '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/uptime',
            null,
            ['range' => $range],
        );
    }

    // ── Mesaj günlüğü ─────────────────────────────────────────────────────
    //
    // Salt okur. Partner kendi panelinde "gönderdiğimiz her şeyin durumu"
    // ekranını çizsin diye (MESSAGING_UNIFICATION_2026-08-25.md §5.1).
    // Alıcı MASKELİ döner, gövde hiç dönmez — gövde zaten saklanmıyor.

    /** @param array<string,mixed> $query channel, status, class, external_ref, q, from, to, per_page */
    public function listMessages(string $companyExternalId, array $query = []): array
    {
        return $this->request(
            'GET',
            '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/messages',
            null,
            $query,
        );
    }

    /** Tek mesaj + olay zaman çizelgesi (kuyruk → gitti → açıldı → tıklandı). */
    public function getMessage(string $companyExternalId, string $messageId): array
    {
        return $this->request(
            'GET',
            '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/messages/' . rawurlencode($messageId),
        );
    }

    /** Kanal bazlı özet: gönderilen / teslim / açılan / tıklanan. */
    public function messageSummary(string $companyExternalId, string $range = '7d'): array
    {
        return $this->request(
            'GET',
            '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/message-summary',
            null,
            ['range' => $range],
        );
    }

    // ── Modül yetkisi ─────────────────────────────────────────────────────

    public function listModules(string $companyExternalId): array
    {
        return $this->request('GET', '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/modules');
    }

    /**
     * "Bu müşteri şu modül için ödeme yaptı, kullanabilir."
     *
     * @param array<string, mixed> $input
     */
    public function grantModule(string $companyExternalId, array $input): array
    {
        return $this->request(
            'POST',
            '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/modules',
            $input,
        );
    }

    /** Yalnız partner'ın KENDİ verdiği hakkı geri alır; plan hakkına dokunmaz. */
    public function revokeModule(string $companyExternalId, string $module): array
    {
        return $this->request(
            'DELETE',
            '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/modules/' . rawurlencode($module),
        );
    }

    // ── Kullanıcı ─────────────────────────────────────────────────────────

    /** @param array<string, mixed> $input */
    public function createUser(string $companyExternalId, array $input): array
    {
        return $this->request(
            'POST',
            '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/users',
            $input,
        );
    }

    public function listUsers(string $companyExternalId): array
    {
        return $this->request('GET', '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/users');
    }

    /** Üyeliği kaldırır, kişinin Signalbird hesabını SİLMEZ. */
    public function removeUser(string $companyExternalId, string $userExternalId): array
    {
        return $this->request(
            'DELETE',
            '/v1/partner/companies/' . rawurlencode($companyExternalId)
                . '/users/' . rawurlencode($userExternalId),
        );
    }

    // ── Gömme ─────────────────────────────────────────────────────────────

    /**
     * Panel ekranını partner sayfasına gömmek için kısa ömürlü jeton üretir
     * (120 sn, tek kullanımlık). Anahtar tarayıcıya inmesin diye bu çağrı
     * partner'ın SUNUCUSUNDAN yapılır.
     *
     * @param array<string, mixed> $input
     */
    public function createEmbedToken(string $companyExternalId, array $input): array
    {
        return $this->request(
            'POST',
            '/v1/partner/companies/' . rawurlencode($companyExternalId) . '/embed',
            $input,
        );
    }

    // ── Taşıma ────────────────────────────────────────────────────────────

    /**
     * @param  array<string, mixed>|null  $body
     * @param  array<string, mixed>  $query
     * @return array{ok: bool, status: int, data: mixed, code: ?string, message: ?string}
     */
    private function request(string $method, string $path, ?array $body = null, array $query = []): array
    {
        $raw = $this->transport($method, $path . self::buildQuery($query), $body);

        $status = (int) ($raw['status'] ?? 0);
        $errno = (int) ($raw['errno'] ?? 0);
        $error = $raw['error'] ?? null;

        if ($raw['body'] === null && ($errno !== 0 || $error)) {
            $code = $errno === CURLE_OPERATION_TIMEOUTED ? 'TIMEOUT' : 'NETWORK_ERROR';

            return $this->fail(0, $code, (string) ($error ?: 'network error'), null);
        }

        $text = (string) ($raw['body'] ?? '');
        $data = null;

        if ($text !== '') {
            $decoded = json_decode($text, true);
            $data = json_last_error() === JSON_ERROR_NONE ? $decoded : $text;
        }

        if ($status >= 200 && $status < 300) {
            return ['ok' => true, 'status' => $status, 'data' => $data, 'code' => null, 'message' => null];
        }

        $code = is_array($data) && isset($data['code']) && is_string($data['code']) && $data['code'] !== ''
            ? $data['code']
            : ($status === 422 ? 'VALIDATION_ERROR' : ($status === 401 ? 'API_KEY_INVALID' : "HTTP_{$status}"));
        $message = is_array($data) && isset($data['message']) && is_string($data['message']) && $data['message'] !== ''
            ? $data['message']
            : "HTTP {$status}";

        return $this->fail($status, $code, $message, $data);
    }

    /**
     * Ham HTTP taşıması (cURL). Testler bu metodu ezerek sahte yanıt döner.
     *
     * @param  array<string, mixed>|null  $body
     * @return array{status: int, body: ?string, error: ?string, errno: int}
     */
    protected function transport(string $method, string $path, ?array $body): array
    {
        $headers = [
            'Accept: application/json',
            'Authorization: Bearer ' . $this->apiKey,
        ];

        $options = [
            CURLOPT_CUSTOMREQUEST => $method,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => $this->timeout,
            CURLOPT_CONNECTTIMEOUT => $this->timeout,
        ];

        if ($body !== null) {
            $headers[] = 'Content-Type: application/json';
            $options[CURLOPT_POSTFIELDS] = json_encode($body, JSON_UNESCAPED_UNICODE);
        }

        $options[CURLOPT_HTTPHEADER] = $headers;

        $handle = curl_init($this->baseUrl . $path);
        curl_setopt_array($handle, $options);

        $response = curl_exec($handle);
        $status = (int) curl_getinfo($handle, CURLINFO_HTTP_CODE);
        $errno = curl_errno($handle);
        $error = curl_error($handle);
        curl_close($handle);

        return [
            'status' => $status,
            'body' => $response === false ? null : (string) $response,
            'error' => $error !== '' ? $error : null,
            'errno' => $errno,
        ];
    }

    /** @return array{ok: bool, status: int, data: mixed, code: ?string, message: ?string} */
    private function fail(int $status, string $code, string $message, mixed $data): array
    {
        if ($this->throwOnError) {
            throw SignalbirdException::fromResponse($code, $status, $message, $data);
        }

        return ['ok' => false, 'status' => $status, 'data' => $data, 'code' => $code, 'message' => $message];
    }

    /** @param array<string, mixed> $query */
    private static function buildQuery(array $query): string
    {
        $pairs = [];

        foreach ($query as $key => $value) {
            if ($value === null) {
                continue;
            }

            if (is_array($value)) {
                foreach ($value as $item) {
                    $pairs[] = rawurlencode((string) $key) . '[]=' . rawurlencode((string) $item);
                }

                continue;
            }

            if (is_bool($value)) {
                $value = $value ? 'true' : 'false';
            }

            $pairs[] = rawurlencode((string) $key) . '=' . rawurlencode((string) $value);
        }

        return $pairs === [] ? '' : '?' . implode('&', $pairs);
    }
}
